### PR TITLE
Cleanup and code optimization

### PR DESCRIPTION
**Pull request by BenjaminMichaelis**
Trying to do some general cleanup and move to using some newer language features

### CODE REVIEW

**Comment by patriksvensson** [comment]
@BenjaminMichaelis Great initiative, but we'll need to hold off on this. Some pretty large changes coming to the code base soon.

**Comment by BenjaminMichaelis** [comment]
Oh yay, excited to see what they are!

**Comment by patriksvensson** [comment]
@BenjaminMichaelis No need to close the PR. Just meant that I cannot merge this quite yet.

**Comment by BenjaminMichaelis** [comment]
Ah, my misunderstanding, no rush at all :)

**Comment by BenjaminMichaelis** [comment]
@patriksvensson I rebased onto main since it looks like there were big changes made, so it should be updated to that.

**Comment by BenjaminMichaelis** [comment]
@patriksvensson is this still something that would be worth me getting rebased

File: src/Spectre.Console.Cli/Internal/Composition/Activators.cs [hunk 1]
```diff
@@ -61,10 +61,7 @@ public ReflectionActivator(Type type)
         _constructor = GetGreediestConstructor(type);
         _parameters = new List<ParameterInfo>();
 
-        foreach (var parameter in _constructor.GetParameters())
-        {
-            _parameters.Add(parameter);
-        }
+        _parameters.AddRange(_constructor.GetParameters());
```
Comment (atifaziz) [inline src/Spectre.Console.Cli/Internal/Composition/Activators.cs:64]: Could be simplified further to:

```c#
_parameters = new List<ParameterInfo>(_constructor.GetParameters());
```

In fact, I'd get rid of the list and directly store the array returned by `_constructor.GetParameters()` in `_parameters`.

File: src/Spectre.Console/Enrichment/ProfileEnricher.cs [hunk 2]
```diff
@@ -2,22 +2,22 @@ namespace Spectre.Console.Enrichment;
...
+        new ContinuaEnricher(),
+        new GitHubEnricher(),
+        new GitLabEnricher(),
+        new GoCDEnricher(),
+        new JenkinsEnricher(),
+        new MyGetEnricher(),
+        new TeamCityEnricher(),
+        new TfsEnricher(),
+        new TravisEnricher(),
+    };
```
Comment (atifaziz) [inline src/Spectre.Console/Enrichment/ProfileEnricher.cs:20]: Couldn't this be just a plain array since the list is never updated?

```suggestion
    private static readonly IProfileEnricher[] _defaultEnrichers =
    {
        new AppVeyorEnricher(),
        new BambooEnricher(),
        new BitbucketEnricher(),
        new BitriseEnricher(),
        new ContinuaEnricher(),
        new GitHubEnricher(),
        new GitLabEnricher(),
        new GoCDEnricher(),
        new JenkinsEnricher(),
        new MyGetEnricher(),
        new TeamCityEnricher(),
        new TfsEnricher(),
        new TravisEnricher(),
    };
```

File: src/Spectre.Console/Widgets/Exceptions/TypeNameHelper.cs [hunk 3]
```diff
@@ -4,7 +4,7 @@ internal static class TypeNameHelper
 {
     // from  https://github.com/benaadams/Ben.Demystifier/blob/main/src/Ben.Demystifier/TypeNameHelper.cs
     // which was adapted from https://github.com/aspnet/Common/blob/dev/shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs
-    public static readonly Dictionary<Type, string> BuiltInTypeNames = new Dictionary<Type, string>
+    public static readonly Dictionary<Type, string> BuiltInTypeNames = new()
```
Comment (atifaziz) [inline src/Spectre.Console/Widgets/Exceptions/TypeNameHelper.cs:7]: While the `{ … }` syntax work for dictionary entries, `[…] = …` reads more natural:

```c#
public static readonly Dictionary<Type, string> BuiltInTypeNames = new()
{
    [typeof(void)] = "void",
    [typeof(bool)] = "bool",
    [typeof(byte)] = "byte",
    [typeof(char)] = "char",
    [typeof(decimal)] = "decimal",
    [typeof(double)] = "double",
    [typeof(float)] = "float",
    [typeof(int)] = "int",
    [typeof(long)] = "long",
    [typeof(object)] = "object",
    [typeof(sbyte)] = "sbyte",
    [typeof(short)] = "short",
    [typeof(string)] = "string",
    [typeof(uint)] = "uint",
    [typeof(ulong)] = "ulong",
    [typeof(ushort)] = "ushort",
};
```

File: src/Spectre.Console/Widgets/Exceptions/TypeNameHelper.cs [hunk 4]
```diff
@@ -24,7 +24,7 @@ internal static class TypeNameHelper
         { typeof(ushort), "ushort" },
     };
 
-    public static readonly Dictionary<string, string> FSharpTypeNames = new Dictionary<string, string>
+    public static readonly Dictionary<string, string> FSharpTypeNames = new()
```
Comment (atifaziz) [inline src/Spectre.Console/Widgets/Exceptions/TypeNameHelper.cs:27]: While the `{ … }` syntax work for dictionary entries, `[…] = …` reads more natural:

```c#
public static readonly Dictionary<string, string> FSharpTypeNames = new()
{
    ["Unit"] = "void",
    ["FSharpOption"] = "Option",
    ["FSharpAsync"] = "Async",
    ["FSharpOption`1"] = "Option",
    ["FSharpAsync`1"] = "Async",
};
```

File: src/Spectre.Console/Widgets/Layout/Layout.cs [hunk 5]
```diff
@@ -213,7 +213,7 @@ protected override IEnumerable<Segment> Render(RenderOptions options, int maxWid
         var map = MakeRenderMap(options, maxWidth);
 
         var layoutLines = new List<SegmentLine>();
-        layoutLines.AddRange(Enumerable.Range(0, height).Select(x => new SegmentLine()));
+        layoutLines.AddRange(Enumerable.Range(0, height).Select(_ => new SegmentLine()));
```
Comment (atifaziz) [inline src/Spectre.Console/Widgets/Layout/Layout.cs:216]: or just:

```c#
var layoutLines = Enumerable.Range(0, height).Select(_ => new SegmentLine()).ToList();
```

File: src/Spectre.Console/Widgets/Layout/Layout.cs [hunk 6]
```diff
@@ -213,7 +213,7 @@ protected override IEnumerable<Segment> Render(RenderOptions options, int maxWid
...
 
         foreach (var (region, lines) in map.Values.Select(x => (x.Region, x.Render)))
```
Comment (atifaziz) [inline src/Spectre.Console/Widgets/Layout/Layout.cs:218]: Perhaps the nested `foreach` here can be folded into a single chain:

```c#
        foreach (var ((index, _), line) in
                 map.Values.Select(x => (x.Region.Y, x.Region.Height, x.Render))
                           .SelectMany(r => layoutLines.Skip(r.Y)
                                                       .Take(r.Y + r.Height)
                                                       .Enumerate()
                                                       .Select(x => (x.Index + r.Y, x.Item))
                                                       .Zip(r.Render)))
        {
            layoutLines[index].AddRange(line);
        }
```

Although I think it reads clearer with the query syntax:

```c#
        foreach (var ((index, _), line) in
                 from r in map.Values
                 from e in layoutLines.Skip(r.Region.Y)
                                      .Take(r.Region.Y + r.Region.Height)
                                      .Enumerate()
                                      .Select(x => (x.Index + r.Region.Y, x.Item))
                                      .Zip(r.Render)
                 select e)
        {
            layoutLines[index].AddRange(line);
        }
```

File: test/Spectre.Console.Tests/Unit/EmojiTests.cs [hunk 7]
```diff
@@ -19,10 +19,10 @@ public void Should_Substitute_Emoji_Shortcodes_In_Markdown()
     public void Should_Contain_Predefined_Emojis()
     {
         // Given, When
-        const string result = "Hello " + Emoji.Known.GlobeShowingEuropeAfrica + "!";
+        const string Result = "Hello " + Emoji.Known.GlobeShowingEuropeAfrica + "!";
```
Comment (atifaziz) [inline test/Spectre.Console.Tests/Unit/EmojiTests.cs:22]: Local constants conventionally use camel-casing so this should be reverted to `result`, like it was.

File: test/Spectre.Console.Cli.Tests/Utilities/ModuleInitializerAttribute.cs [hunk 8]
```diff
@@ -4,5 +4,5 @@ namespace System.Runtime.CompilerServices;
 [AttributeUsage(AttributeTargets.Method, Inherited = false)]
 public sealed class ModuleInitializerAttribute : Attribute
 {
-}
+}
```
Comment (atifaziz) [inline test/Spectre.Console.Cli.Tests/Utilities/ModuleInitializerAttribute.cs:7]: Revert unrelated change?

File: test/Spectre.Console.Tests/Utilities/ModuleInitializerAttribute.cs [hunk 9]
```diff
@@ -4,5 +4,5 @@ namespace System.Runtime.CompilerServices;
 [AttributeUsage(AttributeTargets.Method, Inherited = false)]
 public sealed class ModuleInitializerAttribute : Attribute
 {
-}
+}
```
Comment (atifaziz) [inline test/Spectre.Console.Tests/Utilities/ModuleInitializerAttribute.cs:7]: Revert unrelated change?

File: src/Spectre.Console.lutconfig [hunk 10]
```diff
@@ -0,0 +1,6 @@
+<LUTConfig Version="1.0">
```
Comment (atifaziz) [inline src/Spectre.Console.lutconfig:1]: Was this file intentionally added?

**Comment by phil-scott-78** [comment]
Hey, let's hold off on anything here. We, and really most OSS projects, try to hold off on such large PRs that are code cleanups due to the comments that are showing up here. 

It's an overwhelming amount for us to verify and test out, plus some original code styling is also intentional that we want to keep.
